lib/neo4j/spatial.rb

```diff
@@ -1,143 +1,180 @@
+      alias_method :find_geometries_in_bbox, :bbox
 
-      def find_geometries_in_bbox(layer, minx, maxx, miny, maxy)
+      def within_distance(layer, coordinate, distance, execute: true)
         options = {
           layer: layer,
-          minx: minx,
-          maxx: maxx,
-          miny: miny,
-          maxy: maxy
+          coordinate: coordinate,
+          distanceInKm: distance
         }
-        spatial_post('/ext/SpatialPlugin/graphdb/findGeometriesInBBox', options)
+
+        wrap_spatial_procedure('withinDistance', options, execute: execute)
       end
+      alias_method :find_geometries_within_distance, :within_distance
 
-      def find_geometries_within_distance(layer, pointx, pointy, distance)
+      def intersects(layer, geometry, execute: true)
+        options = {layer: layer, geometry: geometry}
+
+        wrap_spatial_procedure('intersects', options, execute: execute)
+      end
+
+      # TODO: figure out what closest is supposed to do...
+      def closest(layer, coordinate, distance = 100, execute: true)
         options = {
           layer: layer,
-          pointX: pointx,
-          pointY: pointy,
+          coordinate: coordinate,
           distanceInKm: distance
         }
-        spatial_post('/ext/SpatialPlugin/graphdb/findGeometriesWithinDistance', options)
+
+        wrap_spatial_procedure('closest', options, execute: execute)
       end
 
-      def create_spatial_index(name, type = nil, lat = nil, lon = nil)
-        options = {
-          name: name,
-          config: {
-            provider: 'spatial',
-            geometry_type: type || 'point',
-            lat: lat || 'lat',
-            lon: lon || 'lon'
-          }
-        }
-        spatial_post('/index/node', options)
+      def import_shapefile_to_layer(layer, file_uri)
+        options = {layer: layer, file_uri: file_uri}
+
+        spatial_procedure('importShapefileToLayer', options)
       end
 
-      def add_node_to_spatial_index(index, node)
-        options = {
-          uri: "/#{get_id(node)}",
-          key: 'k',
-          value: 'v'
-        }
-        spatial_post("/index/node/#{index}", options)
+      protected
+
+      def spatial_procedure(procedure_name, procedure_args)
+        call_params = procedure_args.keys.map { |key| "{#{key}}" }.join(', ')
+        call_query = "spatial.#{procedure_name}(#{call_params}) YIELD node"
+
+        query_ = Query.new(session: self)
+        query_.call(call_query).params(procedure_args)
       end
 
-      private
+      def spatial_procedure_without_node(procedure_name, procedure_args)
+        call_params = procedure_args.keys.map { |key| "{#{key}}" }.join(', ')
+        call_query = "spatial.#{procedure_name}(#{call_params})"
 
-      def spatial_post(path, options)
-        parse_response! Neo4j::Session.current.connection.post("/db/data/#{path}", options).body
+        query_ = Query.new(session: self)
+        query_.call(call_query).params(procedure_args)
       end
```

Comment:
These two methods could probably be dried up, maybe with a third argument of `, with_node = true`